pymongo/auth_aws.py

```diff
@@ -88,7 +96,13 @@ def _authenticate_aws(credentials, sock_info):
                 # SASL complete.
                 break
     except PyMongoAuthAwsError as exc:
+        # Clear the cached credentials if we hit a failure in auth.
+        _set_cached_credentials(None)
```

Comment:
I think perhaps we should add an API to `AwsSaslContext` to handle exceptions, which would internally clear the cache.

Reply:
That would be a bit odd since a new AwsSaslContext is created for each auth whereas the creds are cached globally.

Reply:
It's either that or make the concept of a cache public I think.

Reply:
What's wrong with the current approach?

Reply:
It feels odd to use a private API from another library outside of testing, even if it is a sibling package.

Reply:
I think it's fine to make it "public". pymongo-auth-aws isn't used by anyone except pymongo anyway.